pypi_resource/pipio.py

```diff
@@ -0,0 +1,235 @@
+            finder = self._build_package_finder(
+                options=options,
+                session=session,
+                platform=options.platform,
+                python_versions=python_versions,
+                abi=options.abi,
+                implementation=options.implementation,
+            )
+
+            requirement_set = RequirementSet(
+                require_hashes=options.require_hashes,
+            )
+            self.populate_requirement_set(
+                requirement_set,
+                args,
+                options,
+                finder,
+                session,
+                self.name,
+                None
+            )
+
+            candidates = []
+            for req in requirement_set.requirements.values():
+                # extract from finder.find_requirement
+                all_candidates = finder.find_all_candidates(req.name)
+                candidates.extend(all_candidates)
+
+        self.candidates = candidates
+
+
+def _input_to_download_args(resconfig, destdir=None) -> List[str]:
+    """ Build pip download args from a resource configuration. """
+
+    package_name = resconfig['source']['name']
+    package_version = resconfig['version']['version']
+
+    args = ['--no-deps', ]
+
+    # TODO:
+    # abi
+    # implementation
+    if resconfig['source']['pre_release']:
+        args.append('--pre')
+
+    if resconfig['source'].get('platform'):
+        args.extend(['--platform', resconfig['source']['platform']])
+
+    if resconfig['source']['packaging'] == 'source':
+        args.extend(['--no-binary', ':all:'])
+    elif resconfig['source']['packaging'] == 'binary':
+        args.extend(['--only-binary', ':all:'])
+
+    if resconfig['source']['python_version']:
+        args.extend(['--python-version', resconfig['source']['python_version']])
+
+    if destdir:
+        args.extend(['--dest', destdir])
+
+    if 'index_url' in resconfig['source']['repository']:
+        url, hostname = get_pypi_url(resconfig, kind='index')
+        args.extend(['--index-url', url,
+                     '--trusted-host', hostname])
+
+    args.append(package_name + ('=={}'.format(package_version) if package_version else ''))
+    return args
+
+
+def get_pypi_url(input, mode='in', kind='repository') -> Tuple[str, str]:
+    """ Get a PyPi URL including authorization details. """
+    repocfg = input['source']['repository']
+
+    key = '%s_url' % (kind,)
+    if key in repocfg:
+        url = repocfg[key]    
+    url_parts = urlsplit(url)
+
+    if mode == 'out' or repocfg.get('authenticate', None) == 'always':
+        host_login = '%s:%s@%s' % (
+            repocfg.get('username', ''),
+            repocfg.get('password', ''),
+            url_parts[1]
+        )
+        host_login = host_login.lstrip(':@')
+    else:
+        host_login = url_parts[1]
+
+    url = urlunsplit((url_parts[0], host_login, url_parts[2], url_parts[3], url_parts[4]))
```

Comment:
This ends up stripping the trailing slash off of the URL, which causes `twine` to fail the upload:
```
RedirectDetected: "https://test.pypi.org/legacy" attempted to redirect to "https://test.pypi.org/legacy/" during upload. Aborting...
```
On `out`, the `twine` command passes in the user name and password with the command, so are the credentials in the URL necessary in `out`? If not, can we do this url-mangling only when `authenticate: always` and `mode = 'in'`?